src/python_testing/matter_testing_infrastructure/chip/testing/taglist_and_topology_test.py

```diff
@@ -92,21 +92,26 @@ def find_tree_root(current_id):
 
 def parts_list_cycles(tree_endpoints: list[int], endpoint_dict: dict[int, Any]) -> list[int]:
     """Returns a list of all the endpoints in the tree_endpoints list that contain cycles"""
-    def parts_list_cycle_detect(visited: set, current_id: int) -> bool:
-        if current_id in visited:
-            return True
+    def parts_list_cycle_detect(visited: set, current_id: int, stack: set) -> bool:
```

Comment:
to find loops you generally only need a single stack, not two separate sets.

The issue here seems to be that we only add to visited, but never remove, so we actually detect more than cycles: we detect re-use across nodes. i.e. we detect non-cycles, but still invalid trees like in your example:

```
   0 ───┐
   ↓    ↓
   1    3
   ↳ 4 ↵
```

We should probably just re-name the error or add a comment explaing. The functionality seems on purpose and it does detect an invalid "tree".